### DISCUSSION TITLE
Using YAML over JSON for fixtures in the future

### DISCUSSION THREAD

**Discussion by spencer-tb** [ethereum/execution-spec-tests#603] [Ideas]
I've been considering that we use yaml for our fixtures instead of json more and more over the past few weeks.

1) The main reason being that we can reference external files within the yaml. As Prague includes system contracts with a large code section we can directly reference it within the file, keeping fixtures clean. This will additionally be VERY useful for Verkle when we want to include a very large Pre (and dummy blocks) for testing the transition.
```yaml
pre:
  0x00000000219ab540356cbb839cbe05303d7705fa:
    nonce: "0x01"
    balance: "0x00"
    code: !!include deposits_contract.json # or deposits_contract.yaml
```

**Some other advantages:**

2) Multiple documents can be added in one file with `---`. This means we could have 1 file per fork with all the fixtures etc if we wanted to. Or compact fork release files that reference each test file using the above method.
3) It still works with pydantic (I don't think we'd need to change anything).
4) More human readable, as indentation is more pythonic.
5) Less syntax also could means slightly smaller file sizes, and release artifacts.
6) On the other hand we can add comments to the fixtures programmatically. We could add the test docstring in a readable format so you know what the test fixture is doing immediately. JSON does not allow comments.

**Main disadvantages:**

1) Slower dumping (writing & parsing). Fill time would increase and test execution time too.
2) Some clients would have to update how they parse the fixtures. Extra overhead for them. I think geth would work out of the box.

**Comment by spencer-tb** [comment]
@danceratopz @marioevz @winsvega

**Comment by winsvega** [comment]
I was thinking about this
`code: !!include deposits_contract.json`

But didn't like the idea as the test becomes not self complete. So you might end up having the file, but missing the includes, and versioning hashes become meaningless.


The main adventage of yaml are multiline fields for comments and code. And hashtag comments.


The yaml test implementation would affect all test clients and their tools, we gotta ask what they think.

---

In the current model human readable are the test src files (json yaml py)
And the test json is for machine as it has compiled evm code.

**Reply by spencer-tb** [reply]
"And the test json is for machine as it has compiled evm code."

This is how it should be now that you mention it
